http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpSamDAO.scala

```diff
@@ -1,74 +1,280 @@
-  */
-class HttpSamDAO(val baseSamServiceURL: String)(implicit system: ActorSystem, materializer: ActorMaterializer, executionContext: ExecutionContext) extends SamDAO with LazyLogging {
-  private val samServiceURL = baseSamServiceURL
-
-  val http = Http(system)
-
-  override def getStatus(): Future[StatusCheckResponse] = {
-    val uri = Uri(samServiceURL + "/status")
-    executeSamRequest[StatusCheckResponse](HttpRequest(GET, uri), ignoreError = true) map { statusCheckResponse =>
-      // If it's an OK status, strip out any subsystem information from the Sam response.
-      // Otherwise, return the StatusCheckResponse as-is, with error information.
-      if (statusCheckResponse.ok) statusCheckResponse.copy(systems = Map.empty)
-      else statusCheckResponse
-    }
+package org.broadinstitute.dsde.workbench.leonardo
+package dao
+
+import java.io.ByteArrayInputStream
+import java.util.concurrent.TimeUnit
+
+import _root_.io.circe.{Decoder, Json, KeyDecoder}
+import ca.mrvisser.sealerate
+import cats.effect.Effect
+import cats.effect.implicits._
+import cats.implicits._
+import cats.mtl.ApplicativeAsk
+import com.google.api.client.googleapis.auth.oauth2.GoogleCredential
+import com.google.api.client.googleapis.javanet.GoogleNetHttpTransport
+import com.google.api.client.json.jackson2.JacksonFactory
+import com.google.api.services.plus.PlusScopes
+import com.google.api.services.storage.StorageScopes
+import com.google.auth.oauth2.ServiceAccountCredentials
+import com.google.common.cache.{CacheBuilder, CacheLoader}
+import io.chrisdavenport.log4cats.Logger
+import org.broadinstitute.dsde.workbench.leonardo.dao.HttpSamDAO._
+import org.broadinstitute.dsde.workbench.leonardo.model._
+import org.broadinstitute.dsde.workbench.leonardo.model.google.ClusterName
+import org.broadinstitute.dsde.workbench.model.google.GoogleProject
+import org.broadinstitute.dsde.workbench.model.{TraceId, ValueObject, WorkbenchEmail}
+import org.broadinstitute.dsde.workbench.util.health.Subsystems.Subsystem
+import org.broadinstitute.dsde.workbench.util.health.{StatusCheckResponse, SubsystemStatus, Subsystems}
+import org.http4s.circe.CirceEntityDecoder._
+import org.http4s.client.Client
+import org.http4s.client.dsl.Http4sClientDsl
+import org.http4s.headers.Authorization
+import org.http4s.{AuthScheme, Charset, Credentials, EntityDecoder, Headers, Method, Request, Response, Status, Uri}
+
+import scala.collection.JavaConverters._
+import scala.concurrent.duration._
+import scala.util.control.NoStackTrace
+import org.broadinstitute.dsde.workbench.leonardo.JsonCodec._
+
+class HttpSamDAO[F[_]: Effect](httpClient: Client[F], config: HttpSamDaoConfig)(implicit logger: Logger[F]) extends SamDAO[F] with Http4sClientDsl[F] {
+  private val saScopes = Seq(PlusScopes.USERINFO_EMAIL, PlusScopes.USERINFO_PROFILE, StorageScopes.DEVSTORAGE_READ_ONLY)
+
+  private[leonardo] def petTokenCache(implicit ev: ApplicativeAsk[F, TraceId]) = CacheBuilder.newBuilder()
```

Comment:
Similar question as above.. should `HttpSamDAO` be responsible for the `petTokenCache`, or should that live in `SamAuthProvider` and `HttpSamDAO` takes arbitrary `Authorization` headers.

Reply:
I think the refactoring as is is big enough that I'll feel more comfortable for more refactoring in a separate PR.